From 9dfc9dec85effdbd649d1263208ce3bcedfe3a64 Mon Sep 17 00:00:00 2001
From: Mathew McBride <matt@traverse.com.au>
Date: Tue, 13 Jan 2026 11:39:28 +1100
Subject: [PATCH] bridge: use generic service bridge for generic arm platform

---
 source/scripts/init/c_registration/02_bridge.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/source/scripts/init/c_registration/02_bridge.c b/source/scripts/init/c_registration/02_bridge.c
index 1f5df97..6224db7 100644
--- a/source/scripts/init/c_registration/02_bridge.c
+++ b/source/scripts/init/c_registration/02_bridge.c
@@ -46,7 +46,7 @@ const char* SERVICE_CUSTOM_EVENTS[] = {"wan-start|/etc/utopia/service.d/service_
 #elif defined (_CBR_PRODUCT_REQ_)
 const char* SERVICE_DEFAULT_HANDLER = "/etc/utopia/service.d/service_bridge_tchcbr.sh";
 const char* SERVICE_CUSTOM_EVENTS[] = { NULL };   //has to be commented due to NOT hit for Commscope platform then also added by patches 
-#elif defined (_COSA_BCM_ARM_)  && ! defined (_PLATFORM_IPQ_) && ! defined (_PLATFORM_RASPBERRYPI_) && ! defined(_PLATFORM_BANANAPI_R4_)
+#elif defined (_COSA_BCM_ARM_)  && ! defined (_PLATFORM_IPQ_) && ! defined (_PLATFORM_RASPBERRYPI_) && ! defined(_PLATFORM_BANANAPI_R4_) && !defined (_PLATFORM_GENERICARM_)
 #if defined(_SCXF11BFL_PRODUCT_REQ_)
 const char* SERVICE_DEFAULT_HANDLER = "/etc/utopia/service.d/service_bridge_generic.sh";
 #else
