##########################################
Date: Apr 16, 2024 1:00 PM
From:
Subject: Add HE 2.4GHz 40MHz bandwidth allowed parameter
Source: Comcast
Upstream-Status: Pending
Signed-off-by: Bogdan Bogush <bogdan.bogush@comcast.com>
##########################################
diff -Naur git.orig/source/hostap-2.11/src/ap/ap_config.c git/source/hostap-2.11/src/ap/ap_config.c
--- git.orig/source/hostap-2.11/src/ap/ap_config.c	2025-04-02 12:51:49.200937658 +0100
+++ git/source/hostap-2.11/src/ap/ap_config.c	2025-04-02 12:53:50.570879719 +0100
@@ -294,6 +294,7 @@
 	conf->reg_sub_cli_eirp_psd = -1;
 	conf->reg_def_cli_eirp = -1;
 	conf->mu_onoff = 15;
+	conf->he_2ghz_40mhz_width_allowed = 1;
 #endif /* CONFIG_IEEE80211AX */
 
 	/* The third octet of the country string uses an ASCII space character
diff -Naur git.orig/source/hostap-2.11/src/ap/ap_config.h git/source/hostap-2.11/src/ap/ap_config.h
--- git.orig/source/hostap-2.11/src/ap/ap_config.h	2025-04-02 12:51:49.200937658 +0100
+++ git/source/hostap-2.11/src/ap/ap_config.h	2025-04-02 12:53:50.570879719 +0100
@@ -1267,6 +1267,7 @@
 	 * of the AP’s BSS.
 	 */
 	int reg_def_cli_eirp;
+	u8 he_2ghz_40mhz_width_allowed;
 
 	bool require_he;
 	u8 mu_onoff;
diff -Naur git.orig/source/hostap-2.11/src/ap/ieee802_11_he.c git/source/hostap-2.11/src/ap/ieee802_11_he.c
--- git.orig/source/hostap-2.11/src/ap/ieee802_11_he.c	2025-04-02 12:51:49.200937658 +0100
+++ git/source/hostap-2.11/src/ap/ieee802_11_he.c	2025-04-02 12:53:50.570879719 +0100
@@ -121,6 +121,9 @@
 		break;
 	}
 
+	if (!hapd->iface->conf->he_2ghz_40mhz_width_allowed)
+		he_oper_chwidth &= ~HE_PHYCAP_CHANNEL_WIDTH_SET_40MHZ_IN_2G;
+
 	ie_size += mcs_nss_size + ppet_size;
 
 	*pos++ = WLAN_EID_EXTENSION;
