##########################################
From: NarendraVarma_Dandu@comcast.com
Date: Jan 10, 2025 1:00 AM
Source: Comcast
##########################################

diff --git a/source/hostap-2.11/src/ap/hostapd.c b/source/hostap-2.11/src/ap/hostapd.c
index 48cc535..d6afaeb 100644
--- a/source/hostap-2.11/src/ap/hostapd.c
+++ b/source/hostap-2.11/src/ap/hostapd.c
@@ -667,7 +667,7 @@
  * If the BSS being removed is the first link, the next link becomes the first
  * link.
  */
-static void hostapd_bss_link_deinit(struct hostapd_data *hapd)
+void hostapd_bss_link_deinit(struct hostapd_data *hapd)
 {
 #ifdef CONFIG_IEEE80211BE
 	int i;
