##########################################
Date: Oct 30, 2025 1:00 PM
From: 
Subject: Fixed MLO enable API
Source: Comcast
Upstream-Status:
Signed-off-by: Bogdan Bogush <bogdan_bogush@comcast.com>
##########################################
diff --git a/source/hostap-2.11/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 8cbaf0db8..e3cd9e47c 100644
--- a/source/hostap-2.11/src/ap/ieee802_11.c
+++ b/source/hostap-2.11/src/ap/ieee802_11.c
@@ -4773,9 +4773,11 @@ bool hostapd_is_mld_ap(struct hostapd_data *hapd)
 	if (!hapd->conf->mld_ap)
 		return false;
 
+#ifndef RDK_ONEWIFI
 	if (!hapd->iface || !hapd->iface->interfaces ||
 	    hapd->iface->interfaces->count <= 1)
 		return false;
+#endif /* RDK_ONEWIFI */
 
 	return true;
 }
